document-store/src/main/java/org/hypertrace/core/documentstore/mongo/MongoCollection.java

```diff
@@ -157,7 +156,19 @@ private String decodeKey(String key) {
 
     // Assume its SimpleAndQuery for now
     DBObject ref = new BasicDBObject(map);
-    final DBCursor dbCursor = collection.find(ref);
+    DBCursor cursor = collection.find(ref);
+
+    Integer offset = query.getOffset();
+    if (offset != null && offset >= 0) {
+      cursor = cursor.skip(offset);
+    }
+
+    Integer limit = query.getLimit();
```

Comment:
Offset and limit should've been done AFTER orderBy. The update is done in the cursor, and not passed to the mongo query. It's better to pass the offset, limit and orderBy to Docstore query. This would perform better and you don't need to worry about the order when it's applied. You can separate the 2 issues though. Most urgent is to fix the offset and limit AFTER sorting.